tempesta_fw/http_parser.c

```diff
@@ -446,20 +522,30 @@ enum {
 	TRY_STR_LAMBDA(str, { }, state)
 
 /**
- * Close currently processed header and set it's EOL length value properly
- * according to parser's EOL state.
+ * Helper function that handles new lines. Closes currently processed header
+ * using @__msg_hdr_finish. If no header is processed, it checks for connection
+ * type to distinguish request and response. In case of response being
+ * processed, it converts @msg of type TfwHttpMsg to outer struct of type
+ * TfwHttpResp and checks if the EOL event belongs to status-line processing. In
+ * latter case it updates @s_line with setting it's EOL length.
  */
 static inline int
-__header_close_at_eol(TfwHttpMsg *msg)
+__handle_newline(TfwHttpMsg *msg, unsigned char *p)
 {
 	TfwHttpParser *parser = &msg->parser;
-	/* Skip headers that were not opened */
-	if (msg->parser.hdr.ptr) {
-		/* LF and CRLF are the only valid EOL values */
-		int eolen = 1 + !!(parser->_tmp.eol == 0xda);
-		tfw_str_set_eolen(&parser->hdr, eolen);
-		return tfw_http_msg_hdr_close(msg, parser->_hdr_tag);
+
+	if (parser->hdr.ptr)
+		return __msg_hdr_finish(p);
+
+	if (TFW_CONN_TYPE(msg->conn) & Conn_Srv) {
```

Comment:
Only responses needs to handle `s_line`. Requests doesn't store the first HTTP-request line separately.